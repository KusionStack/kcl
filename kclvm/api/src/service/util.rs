use crate::gpyrpc::ExecProgramArgs;

/// Transform the str with zero value into [`Option<String>`]
#[inline]
pub(crate) fn transform_str_para(para: &str) -> Option<String> {
    if para.is_empty() {
        None
    } else {
        Some(para.to_string())
    }
}

#[inline]
pub(crate) fn transform_exec_para(
    exec_args: &Option<ExecProgramArgs>,
) -> anyhow::Result<kclvm_runner::ExecProgramArgs> {
    Ok(match exec_args {
        Some(exec_args) => {
            let args_json = serde_json::to_string(exec_args)?;
            kclvm_runner::ExecProgramArgs::from_str(args_json.as_str())
        }
        None => kclvm_runner::ExecProgramArgs::default(),
    })
}
