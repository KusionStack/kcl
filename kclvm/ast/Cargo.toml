[package]
name = "kclvm-ast"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
rustc_span = { path = "../3rdparty/rustc_span" }
serde = { version = "1", features = ["derive"] }
serde_json = "1.0"
indexmap = "1.0"
fancy-regex = "0.7.1"

kclvm-span = {path = "../span", version = "0.1.0"}
kclvm-error = {path = "../error", version = "0.1.0"}

[dev-dependencies]
kclvm-parser = {path = "../parser", version = "0.1.0"}
pretty_assertions = "1.2.1"
