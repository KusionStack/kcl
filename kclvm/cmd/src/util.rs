use clap::ArgMatches;

#[inline]
pub(crate) fn strings_from_matches(matches: &ArgMatches, key: &str) -> Option<Vec<String>> {
    matches.values_of(key).map(|files| {
        files
            .into_iter()
            .map(|v| v.to_string())
            .collect::<Vec<String>>()
    })
}

#[inline]
pub(crate) fn string_from_matches(matches: &ArgMatches, key: &str) -> Option<String> {
    matches.value_of(key).map(|v| v.to_string())
}

#[inline]
pub(crate) fn bool_from_matches(matches: &ArgMatches, key: &str) -> Option<bool> {
    let occurrences = matches.occurrences_of(key);
    if occurrences > 0 {
        Some(true)
    } else {
        None
    }
}

#[inline]
pub(crate) fn u32_from_matches(matches: &ArgMatches, key: &str) -> Option<u32> {
    let occurrences = matches.occurrences_of(key);
    if occurrences > 0 {
        Some(occurrences as u32)
    } else {
        None
    }
}
