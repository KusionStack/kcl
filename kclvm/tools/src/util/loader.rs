use std::fs;

use anyhow::{bail, Context, Result};

pub(crate) trait Loader<T> {
    fn load_ast(&self) -> Result<T>;
}

pub(crate) struct LoaderInner {
    content: String,
}

pub(crate) enum LoaderKind {
    YAML,
    JSON,
}

impl LoaderInner {
    fn new_with_file_path(file_path: &str) -> Result<Self> {
        let content = fs::read_to_string(file_path)
            .with_context(|| format!("Failed to Load '{}'", file_path))?;

        Ok(Self { content })
    }

    fn new_with_str(s: &str) -> Result<Self> {
        Ok(Self {
            content: s.to_string(),
        })
    }

    fn get_content(&self) -> &str {
        &self.content
    }
}

pub(crate) struct DataLoader {
    kind: LoaderKind,
    inner: LoaderInner,
}

impl DataLoader {
    pub(crate) fn new_with_file_path(loader_kind: LoaderKind, file_path: &str) -> Result<Self> {
        Ok(Self {
            kind: loader_kind,
            inner: LoaderInner::new_with_file_path(file_path)
                .with_context(|| format!("Failed to Load '{}'", file_path))?,
        })
    }

    pub(crate) fn new_with_str(loader_kind: LoaderKind, s: &str) -> Result<Self> {
        Ok(Self {
            kind: loader_kind,
            inner: LoaderInner::new_with_str(s)
                .with_context(|| format!("Failed to Load '{}'", s))?,
        })
    }

    pub(crate) fn get_data(&self) -> &str {
        &self.inner.get_content()
    }
}

impl Loader<serde_json::Value> for DataLoader {
    fn load_ast(&self) -> Result<serde_json::Value> {
        let v = match self.kind {
            LoaderKind::JSON => serde_json::to_value(&self.get_data())
                .with_context(|| format!("Failed to String '{}' to Json", self.get_data()))?,
            _ => {
                bail!("Failed to String to Json Value")
            }
        };

        Ok(v)
    }
}

impl Loader<serde_yaml::Value> for DataLoader {
    fn load_ast(&self) -> Result<serde_yaml::Value> {
        let v = match self.kind {
            LoaderKind::YAML => serde_yaml::to_value(&self.get_data())
                .with_context(|| format!("Failed to String '{}' to Yaml", self.get_data()))?,
            _ => {
                bail!("Failed to String to Yaml Value")
            }
        };

        Ok(v)
    }
}
