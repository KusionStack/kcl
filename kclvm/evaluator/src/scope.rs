use crate::error as kcl_error;
use indexmap::{IndexMap, IndexSet};
use kclvm_ast::ast;
use kclvm_error::Position;
use kclvm_runtime::{ValueRef, _kclvm_get_fn_ptr_by_name, MAIN_PKG_PATH, PKG_PATH_PREFIX};
use kclvm_sema::{builtin, pkgpath_without_prefix, plugin};

use crate::{EvalResult, Evaluator, GLOBAL_LEVEL};

/// The evaluator scope.
#[derive(Debug, Default)]
pub struct SchemaSelf {
    pub value: ValueRef,
    pub config: ValueRef,
}

/// The evaluator scope.
#[derive(Debug, Default)]
pub struct Scope {
    /// Scalars denotes the expression statement values without attribute.
    pub scalars: Vec<ValueRef>,
    /// schema_scalar_idx denotes whether a schema exists in the scalar list.
    pub schema_scalar_idx: usize,
    /// Scope normal variables
    pub variables: IndexMap<String, ValueRef>,
    /// Scope closures referenced by internal scope.
    pub closures: IndexMap<String, ValueRef>,
    /// Potential arguments in the current scope, such as schema/lambda arguments.
    pub arguments: IndexSet<String>,
    /// Schema self denotes the scope that is belonged to a schema.
    pub schema_self: Option<SchemaSelf>,
}

impl<'ctx> Evaluator<'ctx> {
    /// Init a scope named `pkgpath` with all builtin functions
    pub(crate) fn init_scope(&self, pkgpath: &str) {
        {
            let mut ctx = self.ctx.borrow_mut();
            let pkg_scopes = &mut ctx.pkg_scopes;
            if pkg_scopes.contains_key(pkgpath) {
                return;
            }
            let scopes = vec![Scope::default()];
            pkg_scopes.insert(String::from(pkgpath), scopes);
        }
        let msg = format!("pkgpath {} is not found", pkgpath);
        // Init all global types including schema and rule
        let module_list: &Vec<ast::Module> = if self.program.pkgs.contains_key(pkgpath) {
            self.program.pkgs.get(pkgpath).expect(&msg)
        } else if pkgpath.starts_with(kclvm_runtime::PKG_PATH_PREFIX)
            && self.program.pkgs.contains_key(&pkgpath[1..])
        {
            self.program
                .pkgs
                .get(&pkgpath[1..])
                .expect(kcl_error::INTERNAL_ERROR_MSG)
        } else {
            panic!("pkgpath {} not found", pkgpath);
        };
        for module in module_list {
            for stmt in &module.body {
                let name = match &stmt.node {
                    ast::Stmt::Schema(schema_stmt) => schema_stmt.name.node.clone(),
                    ast::Stmt::Rule(rule_stmt) => rule_stmt.name.node.clone(),
                    _ => "".to_string(),
                };
                if !name.is_empty() {
                    let name = name.as_str();
                    let _var_name = format!("${}.${}", pkgpath_without_prefix!(pkgpath), name);
                    let global_var_ptr = self.undefined_value();
                    self.add_variable(name, global_var_ptr);
                }
            }
        }
        // Init all builtin functions
        for symbol in builtin::BUILTIN_FUNCTION_NAMES {
            let function_name =
                format!("{}_{}", builtin::KCL_BUILTIN_FUNCTION_MANGLE_PREFIX, symbol);
            let function_ptr = _kclvm_get_fn_ptr_by_name(&function_name);
            self.add_variable(symbol, self._function_value_with_ptr(function_ptr));
        }
        self.enter_scope();
    }

    /// Get the scope level
    pub(crate) fn scope_level(&self) -> usize {
        let current_pkgpath = self.current_pkgpath();
        let ctx = self.ctx.borrow();
        let pkg_scopes = &ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get(&current_pkgpath).expect(&msg);
        // Sub the builtin global scope
        scopes.len() - 1
    }

    /// Enter scope
    pub(crate) fn enter_scope(&self) {
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        let scope = Scope::default();
        scopes.push(scope);
    }

    /// Leave scope
    pub(crate) fn leave_scope(&self) {
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        scopes.pop();
    }

    /// Append a scalar value into the scope.
    pub fn add_scalar(&self, scalar: ValueRef, is_schema: bool) {
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let scopes = pkg_scopes
            .get_mut(&current_pkgpath)
            .unwrap_or_else(|| panic!("pkgpath {} is not found", current_pkgpath));
        if let Some(last) = scopes.last_mut() {
            let scalars = &mut last.scalars;
            // TODO: To avoid conflicts, only the last schema scalar expressions are allowed.
            let schema_scalar_idx = &mut last.schema_scalar_idx;
            if is_schema {
                // Remove the last schema scalar.
                if *schema_scalar_idx < scalars.len() {
                    scalars.remove(*schema_scalar_idx);
                }
                // Override the last schema scalar.
                scalars.push(scalar);
                *schema_scalar_idx = scalars.len() - 1;
            } else {
                scalars.push(scalar);
            }
        }
    }

    /// Append a variable into the scope
    pub fn add_variable(&self, name: &str, pointer: ValueRef) {
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        if let Some(last) = scopes.last_mut() {
            let variables = &mut last.variables;
            if !variables.contains_key(name) {
                variables.insert(name.to_string(), pointer);
            }
        }
    }

    /// Store the variable named `name` with `value` from the current scope, return false when not found
    pub fn store_variable_in_current_scope(&self, name: &str, value: ValueRef) -> bool {
        // Find argument name in the scope
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        let index = scopes.len() - 1;
        let variables = &mut scopes[index].variables;
        if let Some(_var) = variables.get(&name.to_string()) {
            variables.insert(name.to_string(), value);
            return true;
        }
        false
    }

    /// Store the variable named `name` with `value` from the scope, return false when not found
    pub fn store_variable(&self, name: &str, value: ValueRef) -> bool {
        // Find argument name in the scope
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        for i in 0..scopes.len() {
            let index = scopes.len() - i - 1;
            let variables = &mut scopes[index].variables;
            if variables.get(&name.to_string()).is_some() {
                variables.insert(name.to_string(), value);
                return true;
            }
        }
        false
    }

    /// Resolve variable in scope, return false when not found.
    #[inline]
    pub fn resolve_variable(&self, name: &str) -> bool {
        self.resolve_variable_level(name).is_some()
    }

    /// Resolve variable level in scope, return None when not found.
    pub fn resolve_variable_level(&self, name: &str) -> Option<usize> {
        // Find argument name in the scope
        let current_pkgpath = self.current_pkgpath();
        let ctx = self.ctx.borrow();
        let pkg_scopes = &ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get(&current_pkgpath).expect(&msg);
        let mut level = None;
        for i in 0..scopes.len() {
            let index = scopes.len() - i - 1;
            let variables = &scopes[index].variables;
            let arguments = &scopes[index].arguments;
            if variables.get(name).is_some() {
                level = Some(index);
                break;
            }
            if arguments.contains(name) {
                level = Some(index);
                break;
            }
        }
        level
    }

    /// Append a variable or update the existed local variable.
    pub fn add_or_update_local_variable(&self, name: &str, value: ValueRef) {
        let current_pkgpath = self.current_pkgpath();
        let is_local_var = self.is_local_var(name);
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        let mut existed = false;
        // Query the variable in all scopes.
        for i in 0..scopes.len() {
            let index = scopes.len() - i - 1;
            let is_argument = scopes[index].arguments.contains(name);
            let variables_mut = &mut scopes[index].variables;
            match variables_mut.get(&name.to_string()) {
                // If the local variable is found, store the new value for the variable.
                // We cannot update rule/lambda/schema arguments because they are read-only.
                Some(_) if index > GLOBAL_LEVEL && !is_local_var && !is_argument => {
                    variables_mut.insert(name.to_string(), value.clone());
                    existed = true;
                }
                _ => {}
            }
        }
        // If not found, alloc a new variable.
        if !existed {
            // Store the value for the variable and add the variable into the current scope.
            if let Some(last) = scopes.last_mut() {
                last.variables.insert(name.to_string(), value);
            }
        }
    }

    /// Append a variable or update the existed variable
    pub fn add_or_update_global_variable(&self, name: &str, value: ValueRef) {
        // Find argument name in the scope
        let current_pkgpath = self.current_pkgpath();
        let mut ctx = self.ctx.borrow_mut();
        let pkg_scopes = &mut ctx.pkg_scopes;
        let msg = format!("pkgpath {} is not found", current_pkgpath);
        let scopes = pkg_scopes.get_mut(&current_pkgpath).expect(&msg);
        let mut existed = false;
        if let Some(last) = scopes.last_mut() {
            let variables = &mut last.variables;
            if variables.get(&name.to_string()).is_some() {
                variables.insert(name.to_string(), value.clone());
                existed = true;
            }
        }
        if !existed {
            if let Some(last) = scopes.last_mut() {
                let variables = &mut last.variables;
                if !variables.contains_key(name) {
                    variables.insert(name.to_string(), value);
                }
            }
        }
    }

    /// Get the variable value named `name` from the scope, return Err when not found
    pub fn get_variable(&self, name: &str) -> EvalResult {
        let current_pkgpath = self.current_pkgpath();
        self.get_variable_in_pkgpath(name, &current_pkgpath)
    }

    /// Get the variable value named `name` from the scope, return Err when not found
    pub fn get_variable_in_schema(&self, name: &str) -> EvalResult {
        let pkgpath = self.current_pkgpath();
        let ctx = self.ctx.borrow();
        let scope = ctx.pkg_scopes.get(&pkgpath).unwrap().last().unwrap();
        let schema_self = scope.schema_self.as_ref().unwrap();
        let schema_value = &schema_self.value;
        if let Some(value) = schema_value.dict_get_value(name) {
            Ok(value)
        } else {
            let current_pkgpath = self.current_pkgpath();
            self.get_variable_in_pkgpath(name, &current_pkgpath)
        }
    }

    /// Get the variable value named `name` from the scope named `pkgpath`, return Err when not found
    pub fn get_variable_in_pkgpath(&self, name: &str, pkgpath: &str) -> EvalResult {
        let ctx = self.ctx.borrow();
        let pkg_scopes = &ctx.pkg_scopes;
        let pkgpath =
            if !pkgpath.starts_with(kclvm_runtime::PKG_PATH_PREFIX) && pkgpath != MAIN_PKG_PATH {
                format!("{}{}", kclvm_runtime::PKG_PATH_PREFIX, pkgpath)
            } else {
                pkgpath.to_string()
            };
        let mut result = Err(anyhow::anyhow!("name '{}' is not defined", name));
        let is_in_schema = self.is_in_schema();
        // System module
        if builtin::STANDARD_SYSTEM_MODULE_NAMES_WITH_AT.contains(&pkgpath.as_str()) {
            let pkgpath = &pkgpath[1..];
            let _mangle_func_name = format!(
                "{}{}_{}",
                builtin::KCL_SYSTEM_MODULE_MANGLE_PREFIX,
                pkgpath_without_prefix!(pkgpath),
                name
            );
            let value = if pkgpath == builtin::system_module::UNITS
                && builtin::system_module::UNITS_FIELD_NAMES.contains(&name)
            {
                let value_float: f64 = kclvm_runtime::f64_unit_value(name);
                let value_int: u64 = kclvm_runtime::u64_unit_value(name);
                if value_int != 1 {
                    self.int_value(value_int as i64)
                } else {
                    self.float_value(value_float)
                }
            } else {
                todo!()
            };
            Ok(value)
        }
        // Plugin pkgpath
        else if pkgpath.starts_with(plugin::PLUGIN_PREFIX_WITH_AT) {
            let _null_fn_ptr = 0;
            let name = format!("{}.{}", &pkgpath[1..], name);
            let _none_value = self.none_value();
            return Ok(ValueRef::func(
                0,
                0,
                self.undefined_value(),
                &name,
                "",
                true,
            ));
        // User pkgpath
        } else {
            // Global or local variables.
            let scopes = pkg_scopes
                .get(&pkgpath)
                .unwrap_or_else(|| panic!("package {} is not found", pkgpath));
            // Scopes 0 is builtin scope, Scopes 1 is the global scope, Scopes 2~ are the local scopes
            let scopes_len = scopes.len();
            for i in 0..scopes_len {
                let index = scopes_len - i - 1;
                let variables = &scopes[index].variables;
                if let Some(var) = variables.get(&name.to_string()) {
                    // Closure vars, 2 denotes the builtin scope and the global scope, here is a closure scope.
                    let value = if i >= 1 && i < scopes_len - 2 {
                        let last_lambda_scope = self.last_lambda_scope();
                        // Local scope variable
                        if index >= last_lambda_scope {
                            var.clone()
                        } else {
                            // Outer lambda closure
                            let _variables = &scopes[last_lambda_scope].variables;
                            let ptr: Option<&ValueRef> = None;
                            // Lambda closure
                            match ptr {
                                Some(closure_map) => closure_map.dict_get_value(name).unwrap(),
                                None => var.clone(),
                            }
                        }
                    } else {
                        var.clone()
                    };
                    result = Ok(value);
                    break;
                }
            }
            match result {
                Ok(_) => result,
                Err(ref err) => {
                    if !is_in_schema {
                        let mut ctx = self.ctx.borrow_mut();
                        let handler = &mut ctx.handler;
                        let pos = Position {
                            filename: self.current_filename(),
                            line: self.current_line(),
                            column: None,
                        };
                        handler.add_compile_error(&err.to_string(), (pos.clone(), pos));
                        handler.abort_if_any_errors()
                    }
                    result
                }
            }
        }
    }

    /// Get closure map in the current inner scope.
    pub(crate) fn get_current_inner_scope_variable_map(&self) -> ValueRef {
        let var_map = {
            let last_lambda_scope = self.last_lambda_scope();
            // Get variable map in the current scope.
            let pkgpath = self.current_pkgpath();
            let pkgpath = if !pkgpath.starts_with(PKG_PATH_PREFIX) && pkgpath != MAIN_PKG_PATH {
                format!("{}{}", PKG_PATH_PREFIX, pkgpath)
            } else {
                pkgpath
            };
            let ctx = self.ctx.borrow();
            let pkg_scopes = &ctx.pkg_scopes;
            let scopes = pkg_scopes
                .get(&pkgpath)
                .unwrap_or_else(|| panic!("package {} is not found", pkgpath));
            let current_scope = scopes.len() - 1;
            // Get last closure map.

            if current_scope >= last_lambda_scope && last_lambda_scope > 0 {
                let _variables = &scopes[last_lambda_scope].variables;
                // todo: lambda closure in the lambda.
                let ptr: Option<ValueRef> = None;
                let var_map = match ptr {
                    Some(ptr) => ptr.clone(),
                    None => self.dict_value(),
                };
                // Get variable map including schema  in the current scope.
                for i in last_lambda_scope..current_scope + 1 {
                    let variables = &scopes
                        .get(i)
                        .expect(kcl_error::INTERNAL_ERROR_MSG)
                        .variables;
                    for (_key, _ptr) in variables {
                        todo!()
                    }
                }
                var_map
            } else {
                self.dict_value()
            }
        };
        // Capture schema `self` closure.
        if self.is_in_schema() {
            todo!()
        }
        var_map
    }

    /// Load value from name.
    pub fn load_value(&self, pkgpath: &str, names: &[&str]) -> EvalResult {
        if names.is_empty() {
            return Err(anyhow::anyhow!("error: read value from empty name"));
        }
        let name = names[0];
        // Get variable from the scope.
        let get = |name: &str| {
            match (
                self.is_in_schema(),
                self.is_in_lambda(),
                self.is_local_var(name),
            ) {
                // Get from local or global scope
                (false, _, _) | (_, _, true) => self.get_variable(name),
                // Get variable from the current schema scope.
                (true, false, false) => self.get_variable_in_schema(name),
                // Get from local scope including lambda arguments, lambda variables,
                // loop variables or global variables.
                (true, true, _) =>
                // Get from local scope including lambda arguments, lambda variables,
                // loop variables or global variables.
                {
                    match self.resolve_variable_level(name) {
                        // Closure variable or local variables
                        Some(level) if level > GLOBAL_LEVEL => self.get_variable(name),
                        // Schema closure or global variables
                        _ => self.get_variable_in_schema(name),
                    }
                }
            }
        };
        if names.len() == 1 {
            get(name)
        } else {
            let mut value = if pkgpath.is_empty() {
                get(name)
            } else {
                self.ok_result()
            }
            .expect(kcl_error::INTERNAL_ERROR_MSG);
            for i in 0..names.len() - 1 {
                let attr = names[i + 1];
                if i == 0 && !pkgpath.is_empty() {
                    value = self
                        .get_variable_in_pkgpath(attr, pkgpath)
                        .expect(kcl_error::INTERNAL_ERROR_MSG)
                } else {
                    value = value.load_attr(attr)
                }
            }
            Ok(value)
        }
    }
}
