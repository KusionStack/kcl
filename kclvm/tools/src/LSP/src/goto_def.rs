//! GotoDefinition for KCL
//! Github Issue: https://github.com/kcl-lang/kcl/issues/476
//! Now supports goto definition for the following situation:
//! + variable
//! + schema definition
//! + mixin definition
//! + schema attr
//! + attr type

use indexmap::{IndexMap, IndexSet};
use kclvm_ast::pos::{ContainsPos, GetPos};

use kclvm_ast::ast::{Expr, Identifier, ImportStmt, Node, Program, Stmt};
use kclvm_compiler::pkgpath_without_prefix;
use kclvm_error::Position as KCLPos;

use kclvm_sema::resolver::scope::{ProgramScope, Scope, ScopeObject};
use kclvm_sema::ty::{DictType, SchemaType};
use lsp_types::{GotoDefinitionResponse, Url};
use lsp_types::{Location, Range};
use std::cell::RefCell;
use std::path::Path;
use std::rc::Rc;

use crate::to_lsp::lsp_pos;
use crate::util::{
    fix_missing_identifier, get_pkg_scope, get_pos_from_real_path, get_real_path_from_external,
    inner_most_expr_in_stmt,
};

// Navigates to the definition of an identifier.
pub(crate) fn goto_definition(
    program: &Program,
    kcl_pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<lsp_types::GotoDefinitionResponse> {
    match program.pos_to_stmt(kcl_pos) {
        Some(node) => match node.node {
            Stmt::Import(stmt) => goto_def_for_import(&stmt, kcl_pos, prog_scope, program),
            _ => match find_def(node.clone(), kcl_pos, prog_scope) {
                Some(def) => positions_to_goto_def_resp(&def.get_positions()),
                None => None,
            },
        },
        None => None,
    }
}

// Todo: fix ConfigExpr
// ```kcl
// schema Person:
//     name: str
//     data: Data

// schema Data:
//     id: int

// person = Person {
//     data.id = 1
//     data: {
//         id = 1
//     }
//     data: Data {
//         id = 3
//     }
// }

#[derive(Debug)]
pub enum Definition {
    Object(ScopeObject),
    Scope(Scope),
}

impl Definition {
    pub(crate) fn get_positions(&self) -> IndexSet<(KCLPos, KCLPos)> {
        let mut positions = IndexSet::new();
        match self {
            Definition::Object(obj) => {
                positions.insert((obj.start.clone(), obj.end.clone()));
            }
            Definition::Scope(scope) => match &scope.kind {
                kclvm_sema::resolver::scope::ScopeKind::Package(filenames) => {
                    for file in filenames {
                        let dummy_pos = KCLPos {
                            filename: file.clone(),
                            line: 1,
                            column: None,
                        };
                        positions.insert((dummy_pos.clone(), dummy_pos));
                    }
                }
                _ => {
                    positions.insert((scope.start.clone(), scope.end.clone()));
                }
            },
        }
        positions
    }
}

pub(crate) fn find_def(
    node: Node<Stmt>,
    kcl_pos: &KCLPos,
    prog_scope: &ProgramScope,
) -> Option<Definition> {
    fn pre_process_identifier(id: Node<Identifier>, pos: &KCLPos) -> Identifier {
        if !id.contains_pos(pos) && id.node.names.is_empty() {
            return id.node.clone();
        }

        let mut id = id.node.clone();
        let mut names = vec![];
        for name in id.names {
            names.push(name.clone());
            if name.contains_pos(pos) {
                break;
            }
        }
        id.names = fix_missing_identifier(&names);
        if !id.pkgpath.is_empty() {
            id.names[0].node = pkgpath_without_prefix!(id.pkgpath);
        }
        id
    }

    let (inner_expr, parent) = inner_most_expr_in_stmt(&node.node, kcl_pos, None);
    if let Some(expr) = inner_expr {
        if let Expr::Identifier(id) = expr.node {
            let id_node = Node::node_with_pos(
                id.clone(),
                (
                    expr.filename,
                    expr.line,
                    expr.column,
                    expr.end_line,
                    expr.end_column,
                ),
            );
            let id = pre_process_identifier(id_node, kcl_pos);
            match parent {
                Some(schema_expr) => {
                    if let Expr::Schema(schema_expr) = schema_expr.node {
                        let schema_def =
                            find_def(node, &schema_expr.name.get_end_pos(), prog_scope);
                        if let Some(schema) = schema_def {
                            match schema {
                                Definition::Object(obj) => {
                                    let schema_type = obj.ty.into_schema_type();
                                    return find_attr_in_schema(
                                        &schema_type,
                                        &id.names,
                                        &prog_scope.scope_map,
                                    );
                                }
                                Definition::Scope(_) => {
                                    //todo
                                }
                            }
                        }
                    }
                }
                None => {
                    if let Some(inner_most_scope) = prog_scope.inner_most_scope(&kcl_pos) {
                        return resolve_var(&id.names, &inner_most_scope, &prog_scope.scope_map);
                    }
                }
            }
        }
    }
    None
}

/// Similar to vars.rs/resolver_var, find a ScopeObj corresponding to the definition of identifier
pub(crate) fn resolve_var(
    node_names: &[Node<String>],
    current_scope: &Scope,
    scope_map: &IndexMap<String, Rc<RefCell<Scope>>>,
) -> Option<Definition> {
    let names = node_names
        .iter()
        .map(|node| node.node.clone())
        .collect::<Vec<String>>();
    match names.len() {
        0 => None,
        1 => {
            let name = names[0].clone();
            match current_scope.lookup(&name) {
                Some(obj) => match obj.borrow().kind {
                    kclvm_sema::resolver::scope::ScopeObjectKind::Module(_) => {
                        match scope_map.get(&name) {
                            Some(scope) => Some(Definition::Scope(scope.borrow().clone())),
                            None => None,
                        }
                    }
                    _ => Some(Definition::Object(obj.borrow().clone())),
                },
                None => None,
            }
        }
        _ => {
            let name = names[0].clone();
            match current_scope.lookup(&name) {
                Some(obj) => {
                    match &obj.borrow().ty.kind {
                        kclvm_sema::ty::TypeKind::Schema(schema_type) => {
                            find_attr_in_schema(schema_type, &node_names[1..], scope_map)
                        }
                        kclvm_sema::ty::TypeKind::Module(module_ty) => {
                            match scope_map.get(&pkgpath_without_prefix!(module_ty.pkgpath)) {
                                Some(scope) => {
                                    return resolve_var(
                                        &node_names[1..],
                                        &scope.borrow(),
                                        scope_map,
                                    );
                                }
                                None => None,
                            }
                        }
                        kclvm_sema::ty::TypeKind::Dict(DictType { attrs: _, .. }) => {
                            // Todo: find key def in dict
                            None
                        }
                        _ => None,
                    }
                }
                None => None,
            }
        }
    }
}

pub fn find_attr_in_schema(
    schema_type: &SchemaType,
    names: &[Node<String>],
    scope_map: &IndexMap<String, Rc<RefCell<Scope>>>,
) -> Option<Definition> {
    let schema_pkg_scope = get_pkg_scope(&schema_type.pkgpath, scope_map);
    let names = if schema_type.pkgpath.is_empty() {
        &names[1..]
    } else {
        names
    };
    for child in &schema_pkg_scope.children {
        let child_scope = child.borrow();
        if let kclvm_sema::resolver::scope::ScopeKind::Schema(schema_name) = &child_scope.kind {
            if schema_name == &schema_type.name {
                return resolve_var(&names, &child_scope, scope_map);
            }
        }
    }
    None
}

// Convert kcl position to GotoDefinitionResponse. This function will convert to
// None, Scalar or Array according to the number of positions
fn positions_to_goto_def_resp(
    positions: &IndexSet<(KCLPos, KCLPos)>,
) -> Option<GotoDefinitionResponse> {
    match positions.len() {
        0 => None,
        1 => {
            let (start, end) = positions.iter().next().unwrap().clone();
            Some(lsp_types::GotoDefinitionResponse::Scalar(Location {
                uri: Url::from_file_path(start.filename.clone()).unwrap(),
                range: Range {
                    start: lsp_pos(&start),
                    end: lsp_pos(&end),
                },
            }))
        }
        _ => {
            let mut res = vec![];
            for (start, end) in positions {
                res.push(Location {
                    uri: Url::from_file_path(start.filename.clone()).unwrap(),
                    range: Range {
                        start: lsp_pos(start),
                        end: lsp_pos(end),
                    },
                })
            }
            Some(lsp_types::GotoDefinitionResponse::Array(res))
        }
    }
}

fn goto_def_for_import(
    stmt: &ImportStmt,
    _kcl_pos: &KCLPos,
    _prog_scope: &ProgramScope,
    program: &Program,
) -> Option<GotoDefinitionResponse> {
    let pkgpath = &stmt.path;
    let mut real_path =
        Path::new(&program.root).join(pkgpath.replace('.', &std::path::MAIN_SEPARATOR.to_string()));
    let mut positions = get_pos_from_real_path(&real_path);

    if positions.is_empty() && !real_path.exists() {
        real_path =
            get_real_path_from_external(&stmt.pkg_name, pkgpath, program.root.clone().into());
    }

    positions = get_pos_from_real_path(&real_path);

    positions_to_goto_def_resp(&positions)
}

pub(crate) fn get_identifier_last_name(id: &Identifier) -> String {
    match id.names.len() {
        0 => "".to_string(),
        1 => id.names[0].node.clone(),
        _ => {
            if id.names.last().unwrap().node == *"" {
                // MissingExpr
                id.names.get(id.names.len() - 2).unwrap().node.clone()
            } else {
                id.names.last().unwrap().node.clone()
            }
        }
    }
}
