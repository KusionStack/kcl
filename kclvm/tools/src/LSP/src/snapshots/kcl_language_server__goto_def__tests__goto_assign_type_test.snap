---
source: tools/src/LSP/src/goto_def.rs
expression: "format!(\"{:?}\",\n    {\n        let (file, _program, _, gs) =\n        compile_test_file(\"src/test_data/goto_def_test/goto_assign_type_test/goto_assign_type_test.k\");\n        let pos = KCLPos\n        { filename : file.clone(), line : 5, column : Some(17), }; let res =\n        goto_def(& pos, & gs); fmt_resp(& res)\n    })"
---
"path: \"src/test_data/goto_def_test/goto_assign_type_test/goto_assign_type_test.k\", range: Range { start: Position { line: 0, character: 7 }, end: Position { line: 0, character: 15 } }"
